Source/WebCore/platform/CaretAnimation.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+
+#include "Timer.h"
+#include "wtf/Forward.h"
+#include "wtf/Seconds.h"
+#include "wtf/StdLibExtras.h"
+#include <wtf/FastMalloc.h>
+#include <wtf/MonotonicTime.h>
+
+namespace WebCore {
+
+class CaretAnimation;
+
+class CaretAnimationClient {
+public:
+    virtual ~CaretAnimationClient() = default;
+
+    virtual void caretAnimationDidUpdate(CaretAnimation&) { }
+    virtual void caretAnimationWillStart(CaretAnimation&) { }
+    virtual void caretAnimationDidEnd(CaretAnimation&) { }
+};
+
+class CaretAnimation {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    enum class Type {
+        Simple
```

Comment:
Why is this Type thing needed if we're going to use an inheritance?

Reply:
@rniwa This is modeled after `ScrollAnimation`, which also has this `Type` enum. It is needed so that the `is` function can work properly with the polymorphism.

Reply:
You can do that with virtual functions too.

Reply:
Please use a virtual function to do this.

Reply:
how would this be done with virtual functions? Should I have a virtual function that returns a `Type` and use that instead of dynamicDowncast / is ?

Reply:
For starters, it's weird that FrameSelection has to downcast just to call `caretPaint`. A better design would have been to have a virtual caretPaint on the interface object / class itself.

Reply:
The way virtual function based type check works is that there will be a virtual function that returns Type or alternatively a bunch of isX where X is the name/type of a subclass.

Reply:
Or we figure out why the caller needs to check the "type" and make a virtual function that answers the real question it is asking.